autoroutes.pyx

```diff
@@ -47,16 +65,51 @@ cdef class Edge:
             str rest = self.pattern[prefix_len:]
         new_child.connect(self.child, rest)
         self.child = new_child
+        new_child.compile()
         self.pattern = self.pattern[:prefix_len]
 
+    cdef Node join(self, str path):
+        cdef:
+            unsigned int local_index, candidate_index
+            Edge candidate = Edge(path, None)
+        candidate.compile()
+        local_index, candidate_index = self.compare(candidate)
+        del candidate
+        if not local_index:
+            return None
+        if local_index < self.pattern_len:
+            self.branch_at(local_index)
+        if candidate_index < len(path):
+            return self.child.insert(path[candidate_index:])
+        return self.child
+
+    cpdef tuple compare(self, Edge other):
+        cdef unsigned int common_len
+        if self.pattern_prefix_len and other.pattern_prefix_len:
+            common_len = common_root_len(self.pattern_prefix, other.pattern_prefix)
+            if not common_len:  # Nothing common.
+                return 0, 0
+            # At least one prefix is not finished, no need to compare further.
+            elif common_len < self.pattern_prefix_len or common_len < other.pattern_prefix_len:
+                return common_len, common_len
+        elif self.pattern_prefix_len or other.pattern_prefix_len:
+            return 0, 0
+        # We now know prefix are either none or equal.
+        if not self.match_type or self.match_type == MATCH_REGEX or self.match_type != other.match_type:
+            return self.pattern_prefix_len, other.pattern_prefix_len
+        # We now know match types are mergeable, let's see if we should also deal with suffix.
+        if self.pattern_suffix and other.pattern_suffix:
+            common_len = common_root_len(self.pattern_suffix, other.pattern_suffix)
+            if common_len:
+                return self.pattern_end+common_len+1, other.pattern_end+common_len+1
```

Comment:
Add spaces between `+` signs? (not sure it applies with cython)

Reply:
Correct.
That addition was in a slice before this commit, thus without space.